Draupnir 2.0.0 (#3941)
spantaleev/matrix-docker-ansible-deploy: docs/configuring-playbook-bot-draupnir.md

The playbook can install and configure the [Draupnir](https://github.com/the-draupnir-project/Draupnir) moderation bot for you.

See the project's [documentation](https://the-draupnir-project.github.io/draupnir-documentation/) to learn what it does and why it might be useful to you.

This documentation page is about installing Draupnir in bot mode. As an alternative, you can run a multi-instance Draupnir deployment by installing [Draupnir in appservice mode](./configuring-playbook-appservice-draupnir-for-all.md) (called Draupnir-for-all) instead.

If your migrating from Mjolnir skip to [this section](#migrating-from-mjolnir-only-required-if-migrating).

## Prerequisites

### Create a management room

Using your own account, create a new invite only room that you will use to manage the bot. This is the room where you will see the status of the bot and where you will send commands to the bot, such as the command to ban a user from another room. Anyone in this room can control the bot so it is important that you only invite trusted users to this room.

> If you make the management room encrypted (E2EE), then you need to enable Native E2EE (see [below](###Native-E2EE-Support)).

Reviewer: ```suggestion
If you make the management room encrypted (E2EE), then you need to enable the native E2EE support (see [below](#native-e2ee-support)).
```

I'd suggest to check whether the internal link is not broken when changing it :wink:

# Enable Draupnir
matrix_bot_draupnir_enabled: true

# Uncomment and adjust this part if you'd like to use a username different than the default
# matrix_bot_draupnir_login: bot.draupnir

# Generate a strong password for the bot. You can create one with a command like `pwgen -s 64 1`.
matrix_bot_draupnir_password: PASSWORD_FOR_THE_BOT

matrix_bot_draupnir_management_room: "MANAGEMENT_ROOM_ID_HERE"
```

## End-to-End Encryption support

Decide whether you want to support having an Encrypted management room or not. Draupnir can still protect encrypted rooms without encryption support enabled.

Refer to Draupnir's [Documentation](https://the-draupnir-project.github.io/draupnir-documentation/moderator/managing-protected-rooms#protecting-encrypted-rooms) for more information on why you might or might not care about encryption support for protected rooms.

> **Note**: Draupnir does not support running with Pantalaimon as it would break all workflows that involve answering prompts with reactions.

Reviewer: This might as well to be moved above `Adjusting the playbook configuration`.

### Native E2EE Support

To enable the native E2EE support, you need to obtain an access token for Draupnir.

Note that Rust Crypto requires a clean access token that has not touched E2EE so curl is recommended as a method to obtain it. **The access token obtained via Element Web does not work with it**. Refer to the documentation on [how to obtain an access token via curl](obtaining-access-tokens.md#obtain-an-access-token-via-curl).

> #### Adjust the playbook configuration to enable Native E2EE

Reviewer: ```suggestion
To enable the native E2EE support, add the following configuration to your `vars.yml` file:
```

```yaml

> #Enables Native E2EE Support in Draupnir

Reviewer: ```suggestion
# Enables the native E2EE Support
```

matrix_bot_draupnir_enable_experimental_rust_crypto: true
```

> ## Make sure the account is free from rate limiting

Reviewer: ```suggestion
### Make sure the account is free from rate limiting
```

If your homeserver's implementation is Synapse, you will need to prevent it from rate limiting the bot's account. **This is a heavily recomended step. If you do not configure it, Draupnir performance will be degraded.**

This can be done using Synapse's [Admin APIs](https://element-hq.github.io/synapse/latest/admin_api/user_admin_api.html#override-ratelimiting-for-users). They can be accessed both externally and internally.

> To expose the APIs publicly, add the following configuration to your `inventory/host_vars/matrix.example.com/vars.yml` file:

Reviewer: ```suggestion
To expose the APIs publicly, add the following configuration to your `vars.yml` file:
```

To discharge rate limiting, run the following command on systems that ship curl (note that it does not work on outdated Windows 10). Even if the APIs are not exposed to the internet, you should still be able to run the command on the homeserver locally. Before running it, make sure to replace `@bot.draupnir:example.com` with the MXID of your Draupnir:

```sh
curl --header "Authorization: Bearer <access_token>" -X POST https://matrix.example.com/_synapse/admin/v1/users/@bot.draupnir:example.com/override_ratelimit
```

### Obtain an access token

Manual access to Synapse's Admin API requires an access token. Refer to the documentation on [how to obtain an access token](obtaining-access-tokens.md).

⚠️ **Warning**: Access tokens are sensitive information. Do not include them in any bug reports, messages, or logs. Do not share the access token with anyone.

> ## Abuse Reports

Reviewer: ```suggestion
### Abuse Reports
```

Draupnir can receive reports in the management room.

> The bot can intercept the report API endpoint of the client-server API, which requires integration with the reverse proxy in front of the homeserver. If you are using traefik, this playbook can set this up for you:

Reviewer: ```suggestion
The bot can intercept the report API endpoint of the client-server API, which requires integration with the reverse proxy in front of the homeserver. If you are using Traefik, this playbook can set this up for you:
```

```yaml
matrix_bot_draupnir_abuse_reporting_enabled: true
```

<!--
NOTE: this is unsupported by the playbook due to the admin API being inaccessible from containers currently.

The other method polls an Synapse Admin API endpoint, hence it is available only if using Synapse and if the Draupnir user is an admin (see [above](#register-the-bot-account)). To enable it, set `pollReports: true` on `vars.yml` file as below. 
-->

> ## Extending the configuration

Reviewer: ```suggestion
### Extending the configuration
```

Decide whether you want Draupnir to be capable of operating in end-to-end encrypted (E2EE) rooms. This includes the management room and the moderated rooms.

To use Native E2EE support see [this section](### Native E2EE Support)

Rust Crypto is experimental but is considerably more stable than Pantalaimon support and is improving over time with improvements to the element fork of matrix-bot-sdk.

Using [Pantalaimon](configuring-playbook-pantalaimon.md) for E2EE support is unsupported by Draupnir as it breaks core parts of the workflow. Pantalaimon is explicitly unsupported.

### Native E2EE Support

To use Native E2EE support you need to use access token based authentication from [below](####Configuration-without-E2EE-support-or-Native-E2EE).

You also need to activate the support via setting `matrix_bot_draupnir_enable_experimental_rust_crypto` to `true`

> ⚠️ **Warning**: The access token used for authentication must be obtained via means like curl. Using an access token from Element does not work with Rust Crypto. The access token in question can not have touched E2EE before use with Draupnir.

Reviewer: Also I assume the note about the Rust Crypto would be worth mentioning in the section `### Obtain an access token` above.

> #### Configuring Draupnir to use Pantalaimon (Unsupported by upstream.)

Author: The support is currently not removed on either side. It is simply just a bad idea but remains working for now but untested in upstream.

matrix_bot_draupnir_raw_homeserver_url: "{{ matrix_addons_homeserver_client_api_url }}"
```

> #### Configuration without Pantalaimon

Reviewer: I could not follow this change. Is "Configuration without Pantalaimon" either "Configuration with native E2EE support" or "Configuration without native E2EE support" ?

Author: This is one of those changes that must have slipped thru the cracks of not saving in vscode. I have this reworked on my desktop at home i do remember. But i will recreate it on my laptop as im away from home.

matrix_bot_draupnir_abuse_reporting_enabled: true
```

<!-- The other method polls an Synapse Admin API endpoint, hence it is available only if using Synapse and if the Draupnir user is an admin (see [above](#register-the-bot-account)). To enable it, set `pollReports: true` on `vars.yml` file as below. (This is unsupported by the playbook due to the admin API being inaccessible from containers currently.) -->

### Extending the configuration

You can configure additional options by adding the `matrix_bot_draupnir_configuration_extension_yaml` variable.

> For example, to change Draupnir's `acceptInvitesFromSpace` option to `!example:example.org`, add the following configuration to your `vars.yml` file:

Reviewer: ```suggestion
For example, to change Draupnir's `acceptInvitesFromSpace` option to `!qporfwt:example.com`, add the following configuration to your `vars.yml` file:
```

Author: As room IDs are literally free to be whatever you want i dont see the purpose of this example value change. The TLD that might make sense. Readable room IDs with clear meanings are common in communities where users run modified Synapse that allows this.

Reviewer: Please check https://github.com/spantaleev/matrix-docker-ansible-deploy/issues/3654 for the context.

  #
  # If you need something more special, you can take full control by
  # completely redefining `matrix_bot_draupnir_configuration_yaml`.

>   acceptInvitesFromSpace: "!example:example.org"

Reviewer: ```suggestion
  acceptInvitesFromSpace: "!qporfwt:example.com"
```

**Notes**:

- The `ensure-matrix-users-created` playbook tag makes the playbook automatically create the bot's user account.

> 

Author: The command does absolutely nothing for Draupnir as it's not on the list of services that get users automatically created. Or well if it's somehow part I haven't seen it in the list I looked at in group_vars

This could in theory be part of a separate PR but was included as it's part of the work done on 2.0.0 for mdad.

> 

Reviewer: Please see the comment above: https://github.com/spantaleev/matrix-docker-ansible-deploy/pull/3941#discussion_r1917841515
